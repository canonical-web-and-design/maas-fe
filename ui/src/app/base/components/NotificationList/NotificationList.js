import { Notification } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import React, { useEffect } from "react";

import {
  messages as messageActions,
  notification as notificationActions,
} from "app/base/actions";
import NotificationGroup from "app/base/components/NotificationGroup";
import messageSelectors from "app/store/message/selectors";
import notificationSelectors from "app/store/notification/selectors";

const generateMessages = (messages, dispatch) =>
  messages.map(({ id, message, status, temporary, type }) => (
    <Notification
      close={() => dispatch(messageActions.remove(id))}
      key={id}
      status={status}
      timeout={temporary && 5000}
      type={type}
    >
      {message}
    </Notification>
  ));

const NotificationList = () => {
  const messages = useSelector(messageSelectors.all);

  const notifications = {
    warnings: {
      items: useSelector(notificationSelectors.warnings),
      type: "caution",
    },
    errors: {
      items: useSelector(notificationSelectors.errors),
      type: "negative",
    },
    success: {
      items: useSelector(notificationSelectors.success),
      type: "positive",
    },
    info: {
      items: useSelector(notificationSelectors.info),
      type: "information",
    },
  };

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(notificationActions.fetch());
  }, [dispatch]);

  return (
    <>
      {Object.keys(notifications).map((group) => {
        const type = notifications[group].type;
        if (notifications[group].items.length > 0) {
          return (
            <NotificationGroup
              key={type}
              type={type}
              notifications={notifications[group].items}
            />
          );
        }
        return null;
      })}
      {generateMessages(messages, dispatch)}
    </>
  );
};

export default NotificationList;
