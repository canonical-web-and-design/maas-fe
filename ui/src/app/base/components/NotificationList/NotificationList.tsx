import { useEffect } from "react";

import { Notification } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import type { Dispatch } from "redux";

import { messages as messageActions } from "app/base/actions";
import NotificationGroup from "app/base/components/NotificationGroup";
import messageSelectors from "app/store/message/selectors";
import type { Message } from "app/store/message/types";
import { actions as notificationActions } from "app/store/notification";
import notificationSelectors from "app/store/notification/selectors";

const generateMessages = (messages: Message[], dispatch: Dispatch) =>
  messages.map(({ id, message, temporary, type }) => (
    <Notification
      close={() => dispatch(messageActions.remove(id))}
      data-test="message"
      key={id}
      timeout={temporary && 5000}
      type={type}
    >
      {message}
    </Notification>
  ));

const NotificationList = (): JSX.Element => {
  const messages = useSelector(messageSelectors.all);

  const notifications = {
    warnings: {
      items: useSelector(notificationSelectors.warnings),
      type: "caution",
    },
    errors: {
      items: useSelector(notificationSelectors.errors),
      type: "negative",
    },
    success: {
      items: useSelector(notificationSelectors.success),
      type: "positive",
    },
    info: {
      items: useSelector(notificationSelectors.info),
      type: "information",
    },
  };

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(notificationActions.fetch());
  }, [dispatch]);

  return (
    <>
      {Object.keys(notifications).map((group) => {
        const type = notifications[group].type;
        if (notifications[group].items.length > 0) {
          return (
            <NotificationGroup
              key={type}
              type={type}
              notifications={notifications[group].items}
            />
          );
        }
        return null;
      })}
      {generateMessages(messages, dispatch)}
    </>
  );
};

export default NotificationList;
