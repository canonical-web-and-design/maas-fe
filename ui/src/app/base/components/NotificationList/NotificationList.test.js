import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import NotificationList from "./NotificationList";

const mockStore = configureStore();

describe("NotificationList", () => {
  let state;

  beforeEach(() => {
    state = {
      messages: { items: [{ id: 1, message: "User deleted" }] }
    };
  });

  it("renders a list of messages", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <NotificationList />
      </Provider>
    );
    expect(wrapper.find("NotificationList")).toMatchSnapshot();
  });

  it("can hide a notification", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <NotificationList title="Settings">content</NotificationList>
      </Provider>
    );
    wrapper
      .find("Notification")
      .props()
      .close();
    expect(store.getActions()).toEqual([
      {
        type: "REMOVE_MESSAGE",
        payload: 1
      }
    ]);
  });
});
