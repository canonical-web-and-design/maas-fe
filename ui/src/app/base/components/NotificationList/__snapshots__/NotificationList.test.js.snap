// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationList renders a list of messages 1`] = `
<NotificationList>
  <NotificationGroup
    key="negative"
    notifications={
      Array [
        Object {
          "category": "error",
          "id": 1,
          "message": "an error",
        },
      ]
    }
    type="negative"
  >
    <div
      className="row p-notification--group"
    >
      <div
        className="p-notification--negative"
      >
        <div
          className="p-notification__toggle"
        >
          <NotificationGroupMessage
            actionHandler={[Function]}
            id={1}
            message="an error"
          >
            <div
              className="p-notification__response u-no-max-width"
            >
              <ul
                className="p-inline-list u-no-margin--bottom"
              >
                <li
                  className="p-inline-list__item"
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "an error",
                      }
                    }
                    data-test="notification-message"
                  />
                </li>
              </ul>
            </div>
          </NotificationGroupMessage>
        </div>
      </div>
    </div>
  </NotificationGroup>
  <Notification
    close={[Function]}
    key="1"
  >
    <div
      className="p-notification"
    >
      <p
        className="p-notification__response"
      >
        User deleted
      </p>
      <button
        aria-label="Close notification"
        className="p-icon--close"
        onClick={[Function]}
      >
        Close
      </button>
    </div>
  </Notification>
</NotificationList>
`;
